File: app/src/main/java/com/sssprog/shoppingliststandalone/ui/BaseMvpActivity.java
package com.sssprog.shoppingliststandalone.ui;

import android.os.Bundle;

import com.sssprog.shoppingliststandalone.mvp.Presenter;
import com.sssprog.shoppingliststandalone.mvp.PresenterHolder;


public abstract class BaseMvpActivity<P extends Presenter> extends BaseActivity {

    private PresenterHolder<P> presenterHolder;

    public P getPresenter() {
        return presenterHolder.getPresenter();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenterHolder = PresenterHolder.createHolder(this);
        presenterHolder.init(savedInstanceState);
        getPresenter().attach(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPresenter().attach(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPresenter().detach();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        presenterHolder.saveState(outState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (isFinishing()) {
            presenterHolder.onDestroy();
        }
    }

}

File: app/src/main/kotlin/com/sssprog/shoppingliststandalone/ui/BaseMvpActivity.kt
package com.sssprog.shoppingliststandalone.ui

import android.os.Bundle
import com.sssprog.mvp.Presenter
import com.sssprog.shoppingliststandalone.mvp.PresenterKotlin
import kotlin.properties.Delegates

public abstract class BaseMvpActivity<P: PresenterKotlin<*>>: BaseActivity() {

    protected var presenter: P by Delegates.notNull()
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        var p: P? = null
//        savedInstanceState.let {
//            p = PresenterCache.getInstance().get(javaClass)
//        }
//        if (p == null) {
//            p = createPresenter()
//            PresenterCache.getInstance().put(javaClass, p!!)
//        }
//        presenter = p!!
//        (presenter as Presenter<in BaseActivity>).attach(this)


        presenter = createPresenter()
        (presenter as Presenter<in BaseActivity>).attach(this)
    }

    abstract fun createPresenter(): P
}
